controller/volume_controller.go

```diff
@@ -3633,17 +3633,21 @@ func (c *VolumeController) processMigration(v *longhorn.Volume, es map[string]*l
 	// the only time there should be more then 1 engines is when we are migrating or upgrading
 	// if there are more then 1 and we no longer have a migration id set we can cleanup the extra engine
 	if v.Spec.MigrationNodeID == "" {
-		if len(es) < 2 {
-			return nil
-		}
-
 		// in the case of a confirmation we need to switch the v.Status.CurrentNodeID to v.Spec.NodeID
 		// so that currentEngine becomes the migration engine
 		if v.Status.CurrentNodeID != v.Spec.NodeID {
 			log.Infof("volume migration complete switching current node id from %v to %v", v.Status.CurrentNodeID, v.Spec.NodeID)
 			v.Status.CurrentNodeID = v.Spec.NodeID
 		}
 
+		if len(es) < 2 {
```

Comment:
Why moving this condition after the above `if v.Status.CurrentNodeID != v.Spec.NodeID {`? This is part of conditions to do any migration check, so it should be at the beginning, right?

Suggest just changing `if len(es) < 2 {` to below.

```go
if len(es) < 2 {
    if v.Status.CurrentMigrationNodeID == "" {
        return nil
	}
    
    return errors.New(....) // return an intended error to let processMigration rerun next time. 
}
```

Reply:
if `v.Status.CurrentMigrationNodeID != ""` but somehow `len(es) < 2`, the new engine or the old engine may crash and Longhorn should be able to handle this rather than returning an error.

Reply:
Exactly, so probably just continue the following logic to ensure https://github.com/longhorn/longhorn-manager/blob/023dd3cbd5009f390608ea9cdd9f31bd397cf8a2/controller/volume_controller.go#L3677 is reached eventually?

Reply:
Modified 
`if len(es) < 2 {` to be `if len(es) < 2 && v.Status.CurrentMigrationNodeID == "" {`
to handle old engine is deleted then volume status updated failed or new engine or the old engine may crash.
The following code flow could update the correct volume status information again.
https://github.com/longhorn/longhorn-manager/blob/023dd3cbd5009f390608ea9cdd9f31bd397cf8a2/controller/volume_controller.go#L3677  Yes, it could be reached eventually if the engine in the es (engine list) is the new one.